### PR TITLE
Security for _field_names field should not override field statistics

### PR DESCRIPTION
**Pull request by jimczi**
In Lucene 8 the statistics for a field (doc_count, sum_doc_count, ...) are
checked and invalid values (v < 0) are rejected. Though for the _field_names
field we hide the statistics of the field if security is enabled since
some terms (field names) may be filtered. However this statistics are never
used, this field is not used for ranking and cannot be used to generate
term vectors. For these reasons this commit restores the original statistics
for the field in order to be compliant with Lucene 8.

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/FieldSubsetReader.java [hunk 1]
```diff
@@ -109,11 +114,13 @@ public CacheHelper getReaderCacheHelper() {
     private final FieldInfos fieldInfos;
     /** An automaton that only accepts authorized fields. */
     private final CharacterRunAutomaton filter;
+    /** {@link Terms} cache with filtered stats for the {@link FieldNamesFieldMapper} field. */
+    private Terms fieldNamesFilterTerms;
```
Comment (jpountz) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/FieldSubsetReader.java]: let's make it final?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/FieldSubsetReader.java [hunk 2]
```diff
@@ -371,37 +375,47 @@ private Terms wrapTerms(Terms terms, String field) {
      * representing fields that should not be visible in this reader.
      */
     class FieldNamesTerms extends FilterTerms {
+        long size = 0;
+        long sumDocFreq;
+        int docCount;
```
Comment (jpountz) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/FieldSubsetReader.java]: can we make them final somehow?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/FieldSubsetReader.java [hunk 3]
```diff
@@ -371,37 +375,47 @@ private Terms wrapTerms(Terms terms, String field) {
...
-
-        @Override
-        public int getDocCount() throws IOException {
-            return -1;
+        private void computeFilteredStats() throws IOException {
+            TermsEnum e = iterator();
+            while (e.next() != null) {
+                size ++;
+                sumDocFreq += e.docFreq();
+                docCount = Math.max(e.docFreq(), docCount);
```
Comment (jpountz) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/FieldSubsetReader.java]: I don't think this is correct... Maybe we should assume docCount = maxDoc.

**Comment by jimczi** [comment]
We discussed with @jpountz and he proposed that we recompute the stats for this field rather than keeping the wrong stats. This should be fast since the number of terms is bounded by the number of fields in the index and the result can be cached per leaf reader. This idea implemented in https://github.com/elastic/elasticsearch/pull/33261/commits/c3c8ca3df2f5f724820ad1794dfe54ed6d4839ae.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/FieldSubsetReader.java (hunk 1 above)
Comment (jimczi) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/FieldSubsetReader.java]: ++

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/FieldSubsetReader.java (hunk 2 above)
Comment (jimczi) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/FieldSubsetReader.java]: +1

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/FieldSubsetReader.java (hunk 3 above)
Comment (jimczi) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/FieldSubsetReader.java]: oups thanks, I changed it to return maxDoc instead

**Comment by jimczi** [comment]
run gradle build tests

**Review by jpountz — APPROVED** [review]

**Pull request merged**